diff --git a/gst/tcp/gstmultihandlesink.c b/gst/tcp/gstmultihandlesink.c
index b7f76dd..c5d1805 100644
--- a/gst/tcp/gstmultihandlesink.c
+++ b/gst/tcp/gstmultihandlesink.c
@@ -540,6 +540,7 @@ gst_multi_handle_sink_client_init (GstMultiHandleClient * client,
   client->new_connection = TRUE;
   client->sync_method = sync_method;
   client->currently_removing = FALSE;
+  client->handle_write = TRUE;
 
   /* update start time */
   client->connect_time = g_get_real_time () * GST_USECOND;
diff --git a/gst/tcp/gstmultihandlesink.h b/gst/tcp/gstmultihandlesink.h
index 01ee54a..fe65abf 100644
--- a/gst/tcp/gstmultihandlesink.h
+++ b/gst/tcp/gstmultihandlesink.h
@@ -172,6 +172,8 @@ typedef struct {
   guint64 avg_queue_size;
   guint64 first_buffer_ts;
   guint64 last_buffer_ts;
+
+  gboolean handle_write;
 } GstMultiHandleClient;
 
 #define CLIENTS_LOCK_INIT(mhsink)       (g_rec_mutex_init(&(mhsink)->clientslock))
diff --git a/gst/tcp/gstmultisocketsink.c b/gst/tcp/gstmultisocketsink.c
index 56d99d8..d3e73dd 100644
--- a/gst/tcp/gstmultisocketsink.c
+++ b/gst/tcp/gstmultisocketsink.c
@@ -451,6 +451,7 @@ gst_multi_socket_sink_init (GstMultiSocketSink * this)
   this->cancellable = g_cancellable_new ();
   this->send_dispatched = DEFAULT_SEND_DISPATCHED;
   this->send_messages = DEFAULT_SEND_MESSAGES;
+  this->read_buffer = NULL;
 }
 
 static void
@@ -664,6 +665,12 @@ gst_multi_socket_sink_handle_client_read (GstMultiSocketSink * sink,
   } while (navail > 0);
   g_clear_error (&err);
 
+  if (ret) {
+    if (sink->read_buffer) {
+      sink->read_buffer(sink, mhclient, mem, maxmem);
+    }
+  }
+
   if (do_event) {
     if (ret) {
       GstBuffer *buf;
diff --git a/gst/tcp/gstmultisocketsink.h b/gst/tcp/gstmultisocketsink.h
index c89844d..acd7459 100644
--- a/gst/tcp/gstmultisocketsink.h
+++ b/gst/tcp/gstmultisocketsink.h
@@ -65,6 +65,9 @@ typedef struct {
 struct _GstMultiSocketSink {
   GstMultiHandleSink element;
 
+  void          (*read_buffer) (GstMultiSocketSink *sink, GstMultiHandleClient *client,
+                                gchar *buf, gssize size);
+
   /*< private >*/
   GMainContext *main_context;
   GCancellable *cancellable;
diff --git a/gst/tcp/gsttcpelements.h b/gst/tcp/gsttcpelements.h
index 4466083..4056973 100644
--- a/gst/tcp/gsttcpelements.h
+++ b/gst/tcp/gsttcpelements.h
@@ -36,6 +36,7 @@ GST_ELEMENT_REGISTER_DECLARE (tcpclientsink);
 GST_ELEMENT_REGISTER_DECLARE (tcpclientsrc);
 GST_ELEMENT_REGISTER_DECLARE (tcpserversink);
 GST_ELEMENT_REGISTER_DECLARE (tcpserversrc);
+GST_ELEMENT_REGISTER_DECLARE (httpsink);
 GST_ELEMENT_REGISTER_DECLARE (multifdsink);
 GST_ELEMENT_REGISTER_DECLARE (multisocketsink);
 
diff --git a/gst/tcp/gsttcpplugin.c b/gst/tcp/gsttcpplugin.c
index 723ae46..b51418c 100644
--- a/gst/tcp/gsttcpplugin.c
+++ b/gst/tcp/gsttcpplugin.c
@@ -33,6 +33,7 @@ plugin_init (GstPlugin * plugin)
   ret |= GST_ELEMENT_REGISTER (tcpclientsrc, plugin);
   ret |= GST_ELEMENT_REGISTER (tcpserversink, plugin);
   ret |= GST_ELEMENT_REGISTER (tcpserversrc, plugin);
+  ret |= GST_ELEMENT_REGISTER (httpsink, plugin);
 
 #ifdef HAVE_SYS_SOCKET_H
   ret |= GST_ELEMENT_REGISTER (multifdsink, plugin);
diff --git a/gst/tcp/meson.build b/gst/tcp/meson.build
index 98b9686..e035e06 100644
--- a/gst/tcp/meson.build
+++ b/gst/tcp/meson.build
@@ -4,6 +4,7 @@ tcp_sources = [
   'gstsocketsrc.c',
   'gsttcpclientsrc.c',
   'gsttcpclientsink.c',
+  'gsthttpsink.c',
   'gsttcpelements.c',
   'gsttcpserversrc.c',
   'gsttcpserversink.c',
diff --git a/scripts/dist-translations.py b/scripts/dist-translations.py
old mode 100755
new mode 100644
diff --git a/scripts/extract-release-date-from-doap-file.py b/scripts/extract-release-date-from-doap-file.py
old mode 100755
new mode 100644
diff --git a/scripts/gen-changelog.py b/scripts/gen-changelog.py
old mode 100755
new mode 100644
diff --git a/scripts/meson-pkg-config-file-fixup.py b/scripts/meson-pkg-config-file-fixup.py
old mode 100755
new mode 100644
diff --git a/scripts/update-orc-dist-files.py b/scripts/update-orc-dist-files.py
old mode 100755
new mode 100644
diff --git a/tests/examples/gl/cocoa/cocoa-videooverlay.m b/tests/examples/gl/cocoa/cocoa-videooverlay.m
old mode 100755
new mode 100644
