diff --git a/ext/hls/gsthlssink.c b/ext/hls/gsthlssink.c
index 72afe030c..6a885e229 100644
--- a/ext/hls/gsthlssink.c
+++ b/ext/hls/gsthlssink.c
@@ -322,8 +322,8 @@ gst_hls_sink_handle_message (GstBin * bin, GstMessage * message)
       /* multifilesink is an internal implementation detail. If applications
        * need a notification, we should probably do our own message */
       GST_DEBUG_OBJECT (bin, "dropping message %" GST_PTR_FORMAT, message);
-      gst_message_unref (message);
-      message = NULL;
+      //gst_message_unref (message);
+      //message = NULL;
       break;
     }
     case GST_MESSAGE_EOS:{
